unit/ctest_polarization.c

```diff
@@ -0,0 +1,424 @@
+#include "gkyl_array.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <gkyl_proj_on_basis.h>
+#include <gkyl_eval_on_nodes.h>
+#include <gkyl_range.h>
+#include <gkyl_rect_decomp.h>
+#include <gkyl_rect_grid.h>
+#include <gkyl_array_ops.h>
+#include <gkyl_gk_geometry.h>
+#include <gkyl_util.h>
+#include <gkyl_array_rio.h>
+#include <acutest.h>
+#include <gkyl_gyrokinetic_pol_density.h>
+
+// Allocate array (filled with zeros).
+static struct gkyl_array*
+mkarr(bool on_gpu, long nc, long size)
+{
+  struct gkyl_array* a;
+  if (on_gpu)
+    a = gkyl_array_cu_dev_new(GKYL_DOUBLE, nc, size);
+  else
+    a = gkyl_array_new(GKYL_DOUBLE, nc, size);
+  return a;
+}
+
+
+void evalFunc1x_1(double t, const double *xn, double* restrict fout, void *ctx)
+{
+  fout[0] = 1.0;
+}
+
+void evalFunc1x_quad(double t, const double *xn, double* restrict fout, void *ctx)
+{
+  fout[0] = -pow(xn[0]-0.5, 2) + 1.0;
+}
+
+void evalFunc2x_quad(double t, const double *xn, double* restrict fout, void *ctx)
+{
+  fout[0] = (-pow(xn[0]-0.5, 2) + 1.0) * (-pow(xn[1]-0.5, 2) + 1.0);
+}
+
+void evalFunc3x_quad(double t, const double *xn, double* restrict fout, void *ctx)
+{
+  fout[0] = (-pow(xn[0]-0.5, 2) + 1.0) * (-pow(xn[1]-0.5, 2) + 1.0) * (-pow(xn[2]-0.5, 2) + 1.0);
+}
+
+void test_1x_flat( bool use_gpu ) 
+{  
+  int cells[] = {8};
+  int poly_order = 1;
+  double lower[] = {0.0}, upper[] = {1.0};
+  int dim = sizeof(lower)/sizeof(lower[0]);
+  // Grids.
+  struct gkyl_rect_grid grid;
+  gkyl_rect_grid_init(&grid, dim, lower, upper, cells);
+
+  // Basis functions.
+  struct gkyl_basis basis;
+  gkyl_cart_modal_serendip(&basis, dim, poly_order);
+
+  int ghost[] = { 1, 1 };
+  struct gkyl_range localRange, localRange_ext; // local, local-ext ranges.
+  gkyl_create_grid_ranges(&grid, ghost, &localRange_ext, &localRange);
+
+  struct gkyl_gyrokinetic_pol_density* npol_op = gkyl_gyrokinetic_pol_density_new(basis, grid, use_gpu);
+  struct gkyl_array *npol_ho = mkarr(false, basis.num_basis, localRange_ext.volume);
+  struct gkyl_array *epsilon_ho = mkarr(false, basis.num_basis, localRange_ext.volume);
+
+  struct gkyl_eval_on_nodes *epsilon_proj = gkyl_eval_on_nodes_new(&grid, &basis,
+    1, evalFunc1x_1, NULL);
+  gkyl_eval_on_nodes_advance(epsilon_proj, 0.0, &localRange, epsilon_ho);
+  gkyl_eval_on_nodes_release(epsilon_proj);
+
+  struct gkyl_array *npol_dev = use_gpu? mkarr(true, basis.num_basis, localRange_ext.volume)
```

Comment:
please change your tests so these _dev's don't appear. Only the _ho arrays need a suffix.

In fact, looking at this logic, I think it's not quire right. You are calling it _dev but if use_gpu=false it is not actually a dev array.